src/pybind/cudamatrix/cu_vector_pybind.cc

```diff
@@ -34,11 +35,30 @@ void pybind_cu_vector(py::module& m) {
         .def("SetZero", &PyClass::SetZero)
         .def("Set", &PyClass::Set, py::arg("value"))
         .def("Add", &PyClass::Add, py::arg("value"))
-        .def("Scale", &PyClass::Scale, py::arg("value"));
+        .def("Scale", &PyClass::Scale, py::arg("value"))
+        .def("__getitem__", [](const PyClass& v, int i) { return v(i); });
+  }
+  {
+    using PyClass = CuVector<float>;
+    py::class_<PyClass, CuVectorBase<float>>(m, "FloatCuVector")
+        .def(py::init<>())
+        .def(py::init<MatrixIndexT, MatrixResizeType>(), py::arg("dim"),
+             py::arg("MatrixResizeType") = kSetZero)
+        .def(py::init<const VectorBase<float>&>(), py::arg("v"))
+        .def("to_dlpack", [](PyClass* v) {
+          // we use the name `to_dlpack` because PyTorch uses the same name
+          return CuVectorToDLPack(v);
```

Comment:
What I was thinking here was that we'd have this DLManagedTensor do an incref on `v` when it is created, and a decref on `v` when it is deleted.  Of course, I was thinking we'd do this for just Vector, and not VectorBase.  The idea is that the Vector itself is the owner of the memory it owns, and deleting the Vector itself is the only valid way of freeing that memory.   The thing I am now sure, about, though, is how we get access to the Python object that we are part of here.  I don't know the invocation.

Reply:
.. of course the other option would be to have the creation of the DLManagedTensor be destructive of the original Vector or Matrix, and to have the DLManagedTensor itself own the data pointer.  But I think it's preferable to have it non-destructive, if we can find a way to incref the python object.  I imagine that these pybind-wrapped classes must somehow 'live inside' a Python object.

Reply:
That is easy.

The returned object is a `py::capsule` (https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L1145)

which is a subclass of `py::handle` (https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L174)

which has `inc_ref` method (https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L190)

The question is why we need to to this in Python ?

In C++, `SubVector` borrows a pointer from an outside object, it is the callers
respponsibility to keep the outside object alive while the `SubVector` is still in use.

In Python, the caller has to ensure the `Matrix` is alive if he/she calls 
`dlpack = matrix.to_dlpack()`.

Reply:
Here is the PyTorch's implementation for `to_dlpack`

https://github.com/pytorch/pytorch/blob/1dfb8752a62de49b1d270aa0c84d05aee07e51db/torch/csrc/Module.cpp#L370
```cpp
PyObject *THPModule_toDLPack(PyObject *_unused, PyObject *data)
{
  HANDLE_TH_ERRORS
  THPUtils_assert(THPVariable_Check(data), "data must be a Tensor");
  DLManagedTensor* dlMTensor = at::toDLPack(THPVariable_Unpack(data));
  return PyCapsule_New(dlMTensor, "dltensor", DLPack_Capsule_Destructor);
  END_HANDLE_TH_ERRORS
}
```

https://github.com/pytorch/pytorch/blob/1dfb8752a62de49b1d270aa0c84d05aee07e51db/aten/src/ATen/DLConvertor.cpp#L162
```cpp
// This function returns a shared_ptr to memory managed DLpack tensor
// constructed out of ATen tensor
DLManagedTensor* toDLPack(const Tensor& src) {
  ATenDLMTensor* atDLMTensor(new ATenDLMTensor);
  atDLMTensor->handle = src;
  atDLMTensor->tensor.manager_ctx = atDLMTensor;
  atDLMTensor->tensor.deleter = &deleter;
  atDLMTensor->tensor.dl_tensor.data = src.data_ptr();
  int64_t device_id = 0;
  if (src.is_cuda()) {
    device_id = src.get_device();
  }
  atDLMTensor->tensor.dl_tensor.ctx = getDLContext(src, device_id);
  atDLMTensor->tensor.dl_tensor.ndim = src.dim();
  atDLMTensor->tensor.dl_tensor.dtype = getDLDataType(src);
  atDLMTensor->tensor.dl_tensor.shape =
      const_cast<int64_t*>(src.sizes().data());
  atDLMTensor->tensor.dl_tensor.strides =
      const_cast<int64_t*>(src.strides().data());
  atDLMTensor->tensor.dl_tensor.byte_offset = 0;
  return &(atDLMTensor->tensor);
}
```

https://github.com/pytorch/pytorch/blob/1dfb8752a62de49b1d270aa0c84d05aee07e51db/aten/src/ATen/DLConvertor.cpp#L151
```cpp
struct ATenDLMTensor {
  Tensor handle;
  DLManagedTensor tensor;
};
```

It does not increase the reference count of the underlying tensor.
The caller has to ensure it is alive when the PyCapsule is in use.

Reply:
I think the thing that's different here is that class Tensor internally has some kind of shared_ptr.  (I've looked into PyTorch code and I think this is true, but it may be renamed to some kind of PyTorch-specific shared_ptr).
So the Tensor itself doesn't have to be kept alive.  Kaldi's memory management is a bit older and simpler, the Matrix itself has to be kept alive if you want the memory to not be deleted, so the only way to ensure this is to incref that Matrix.